[Bug] CI Pipeline Does Not Fail when Vulnerable Dependencies Found #493

VladislavAntonyuk opened the pull request:
<!--
 Hello, and thank you for your interest in contributing to the .NET MAUI Toolkit! 

 Before you submit please check that this work relates to one of the following:
 - Bug fix
    If you haven't yet opened an Issue that reports the bug in detail, provides a reproduction sample, and has been verified + reproduced by a member of the .NET MAUI Toolkit core team, please do that before submitting a Pull Request.
 - Feature/Proposal
    If you haven't yet submitted a Proposal that has been Championed by a .NET MAUI core team member, please instead open a Discussion at https://github.com/communitytoolkit/maui/discussions/new where we can discuss the pros/cons of the feature and its implementation. 
 Any PR submitted that does not fit with the above options will be closed.
 -->

 ### Description of Change ###

Fail pipeline if has any vulnerability

 ### Linked Issues ###
 <!-- Provide links to issues here (#35 will link to issue number 35). Ensure that a GitHub issue was created for your bug/proposal and it has been approved/Championed. -->

 - Fixes #493

 ### PR Checklist ###
 <!--
 Please check all the things you did here and double-check that you got it all, or state why you didn't do something.

 If anything is unclear please do ask :)
 -->
 - [x] Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
 - [x] Has tests (if omitted, state reason in description)
 - [x] Has samples (if omitted, state reason in description)
 - [x] Rebased on top of `main` at time of PR
 - [x] Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui/blob/main/CONTRIBUTING.md#contributing-code---best-practices)
 - [ ] Documentation created or updated: https://github.com/MicrosoftDocs/CommunityToolkit/pull/XYZ <!-- Replace XYZ with your docs PR #. The checkbox will be automatically checked once the docs PR has been merged -->


 ### Additional information ###

 <!-- 
 Please use this to aid the reviewer, this could include stating which platform(s) have been tested
 -->

VladislavAntonyuk commented:
As we have 3 vulnerabilities, how can we move forward with this PR? Some of them were fixed by .NET MAUI team

jfversluis commented:
Oh good one! Why don't we get dependabot alerts on this repo? That should also catch these, right?

VladislavAntonyuk commented:
Added dependabot. I added the flag `CheckDependencies`. we can set it to false to skip checks on each build and set it to trueenable it during the release build.

VladislavAntonyuk commented:
Question №2, should we check Transitive vulnerabilities? we cannot fix them.
Question №3. Should we check it in solution or only src projects? We have vulnerabilities in CommunityToolkit.Maui.UnitTests

brminnick commented on .github/dependabot.yml:
```diff
@@ -0,0 +1,7 @@
+version: 2
+updates:
+- package-ecosystem: nuget
+  directory: "/"
+  schedule:
+    interval: daily
+  open-pull-requests-limit: 10
```
Let's point dependabot to the `/samples` directory so that it checks all of the NuGet packages in `CommunityToolkit.Maui.Sample.sln`, and let's remove the `open-pull-requests-limit`

```suggestion
version: 2
updates:
- package-ecosystem: nuget
  directory: "/samples"
  schedule:
    interval: daily
```

brminnick commented:
Thanks Vlad!

> As we have 3 vulnerabilities, how can we move forward with this PR? Some of them were fixed by .NET MAUI team

The 3 vulnerabilities are in the `*.UnitTest` projects. I agree with your approach of adding/updating the NuGet Packages in the UnitTest csprojs.

> Question №2, should we check Transitive vulnerabilities? we cannot fix them.

We should! We can always add an updated version of the specific NuGet Package that has the vulnerability, like you've done by adding a direct reference to `Newtonsoft.Json`

> Question №3. Should we check it in solution or only src projects? We have vulnerabilities in CommunityToolkit.Maui.UnitTests

I vote to check the entire solution. It's good to ensure we avoid vulnerabilities in every csproj, even if it's a Sample App or a Unit Test.

jfversluis commented:
I agree with everything Brandon said!

VladislavAntonyuk commented:
awesome, it works: https://dev.azure.com/dotnet/CommunityToolkit/_build/results?buildId=74926&view=logs&jobId=792604ca-8f43-5a41-d895-10758edbd758&j=792604ca-8f43-5a41-d895-10758edbd758&t=5abd79d0-3f8c-531b-9955-a83131c96c54

brminnick commented on azure-pipelines.yml:
```diff
@@ -15,7 +15,8 @@ variables:
   XcodeVersion: '13.3.1'
-  RollbackFile: '6.0.312.json'
+  DotNetMauiRollbackFile: '6.0.312.json'
+  CheckDependencies: 'true'
```
Let's rename this variable name to be a question since its a `bool`. 

```suggestion
  ShouldCheckDependencies: 'true'
```

brminnick replied:
I don't know YAML well. Do we need the single quotes?

Or can it be `ShouldCheckDependencies: true`?

brminnick reviewed (commented):
Looking good! Thanks Vlad!

I had just a small bit of feedback around the new environment variable in the pipeline.

brminnick reviewed (approved):
Thanks Vlad!!
